### PR TITLE
feat: add typescript typegen endpoint

### PR DESCRIPTION
**Pull request by soedirgo**
e.g. for current test fixtures:

https://github.com/supabase/postgres-meta/blob/38c325f6eb81fe82806d5c17f4e288df3bc5be8c/test/db/00-init.sql#L1-L53

https://github.com/supabase/postgres-meta/blob/38c325f6eb81fe82806d5c17f4e288df3bc5be8c/test/db/01-memes.sql#L1-L34

Generated types:
```ts
export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json }
  | Json[];

export interface Database {
  public: {
    Tables: {
      users: {
        Row: {
          id: number;
          name: string | null;
          status: "ACTIVE" | "INACTIVE" | null;
        };
        Insert: {
          id?: number;
          name?: string | null;
          status?: "ACTIVE" | "INACTIVE" | null;
        };
        Update: {
          id?: number;
          name?: string | null;
          status?: "ACTIVE" | "INACTIVE" | null;
        };
      };
      todos: {
        Row: {
          id: number;
          details: string | null;
          "user-id": number;
        };
        Insert: {
          id?: number;
          details?: string | null;
          "user-id": number;
        };
        Update: {
          id?: number;
          details?: string | null;
          "user-id"?: number;
        };
      };
      users_audit: {
        Row: {
          id: number;
          user_id: number | null;
          previous_value: Json | null;
          created_at: string | null;
        };
        Insert: {
          id?: number;
          user_id?: number | null;
          previous_value?: Json | null;
          created_at?: string | null;
        };
        Update: {
          id?: number;
          user_id?: number | null;
          previous_value?: Json | null;
          created_at?: string | null;
        };
      };
      category: {
        Row: {
          id: number;
          name: string;
        };
        Insert: {
          id?: number;
          name: string;
        };
        Update: {
          id?: number;
          name?: string;
        };
      };
      memes: {
        Row: {
          name: string;
          category: number | null;
          metadata: Json | null;
          created_at: string;
          id: number;
          status: "new" | "old" | "retired" | null;
        };
        Insert: {
          name: string;
          category?: number | null;
          metadata?: Json | null;
          created_at: string;
          id?: number;
          status?: "new" | "old" | "retired" | null;
        };
        Update: {
          name?: string;
          category?: number | null;
          metadata?: Json | null;
          created_at?: string;
          id?: number;
          status?: "new" | "old" | "retired" | null;
        };
      };
    };
    Functions: {
      add: {
        Args: Record<string, unknown>;
        Returns: number;
      };
    };
  };
}
```

### CODE REVIEW

File: src/server/routes/generators/typescript.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import prettier from 'prettier'
+import { PostgresMeta, PostgresType } from '../../../lib'
+import { DEFAULT_POOL_CONFIG } from '../../constants'
+import { extractRequestForLogging } from '../../utils'
+
+export default async (fastify: FastifyInstance) => {
+  fastify.get<{
+    Headers: { pg: string }
+    Querystring: {
+      excluded_schemas?: string
```
Comment (soedirgo) [inline src/server/routes/generators/typescript.ts:11]: this is for `auth`, `storage`, `graphql`, `realtime` etc. - we should only generate for user's schemas

File: src/server/routes/generators/typescript.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+              )}
+          }
+        }`
+    )}
+}`
+
+    output = prettier.format(output, {
+      parser: 'typescript',
+    })
+    return output
```
Comment (soedirgo) [inline src/server/routes/generators/typescript.ts:159]: `Content-Type: text/plain; charset=utf-8` by default

File: src/server/routes/generators/typescript.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            ${functions
+              .filter(
+                (function_) =>
+                  function_.schema === schema.name && function_.return_type !== 'trigger'
+              )
+              .map(
+                (function_) => `${JSON.stringify(function_.name)}: {
+                  Args: ${(() => {
+                    if (function_.argument_types === '') {
+                      return 'Record<PropertyKey, never>'
```
Comment (soedirgo) [inline src/server/routes/generators/typescript.ts:127]: for functions w/o args

**Review by kiwicopple — APPROVED** [review]
This looks great @soedirgo 

Perhaps just move the templating logic out of the "routes" file 👍

**Comment by soedirgo** [comment]
> Perhaps just move the templating logic out of the "routes" file 👍

Done!

**Pull request merged**